File: app/src/main/java/com/neliry/banancheg/videonotes/mvvm/ThemeViewModel.java
package com.neliry.banancheg.videonotes.mvvm;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import com.neliry.banancheg.videonotes.mvvm.repository.FirebaseDatabaseRepository;
import com.neliry.banancheg.videonotes.mvvm.repository.ThemeRepository;

import java.util.List;

public class ThemeViewModel extends ViewModel {

private MutableLiveData<List<Theme>> themes;
private ThemeRepository repository = new ThemeRepository();

public LiveData<List<Theme>> getThemes() {
    if (themes == null) {
        themes = new MutableLiveData<>();
        loadThemes();
    }
    return themes;
}

@Override
protected void onCleared() {
    repository.removeListener();
}

private void loadThemes() {
    repository.addListener(new FirebaseDatabaseRepository.FirebaseDatabaseRepositoryCallback<Theme>() {
        @Override
        public void onSuccess(List<Theme> result) {
            themes.setValue(result);
        }

        @Override
        public void onError(Exception e) {
            themes.setValue(null);
        }
    });
}
}

File: app/src/main/java/com/neliry/banancheg/videonotes/mvvm/ThemeViewModel.kt
package com.neliry.banancheg.videonotes.mvvm

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.util.Log
import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.neliry.banancheg.videonotes.mvvm.repository.FirebaseDatabaseRepository
import com.neliry.banancheg.videonotes.mvvm.repository.ThemeRepository

class ThemeViewModel : ViewModel() {

    var list = ArrayList<Theme>()
    private var themes: MutableLiveData<List<Theme>>? = null
    private val repository = ThemeRepository()

    fun getThemes(): LiveData<List<Theme>> {
        if (themes == null) {
            themes = MutableLiveData()
            loadThemes()
        }
        return themes!!
    }



     fun loadThemes() {
        repository.databaseReference.addValueEventListener(object: ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onDataChange(dataSnapshot: DataSnapshot) {
                list.clear()
                for (child in dataSnapshot.getChildren()) {
                    val data = child.getValue(repository.getEntityClass());
                    list.add(data!!)
                }
                themes!!.value = list
            }


        })


    }
}
